source/transactions/tests/pin-mongos.yml

```diff
@@ -296,3 +305,180 @@ tests:
           - {_id: 2}
           - {_id: 3}
           - {_id: 4}
+
+  - description: remain pinned after errors within a transaction and commit
+    topology: ["sharded"]
+    clientOptions:
+      useMultipleMongoses: true
+    operations:
+      - name: startTransaction
+        object: session0
+      - name: insertOne
+        object: collection
+        arguments:
+          session: session0
+          document:
+            _id: 3
+        result:
+          insertedId: 3
+      # Enable the fail point only on the Mongos that session0 is pinned to.
+      - name: failPoint
+        session: session0
+        failPoint:
+          configureFailPoint: failCommand
+          mode: { times: 1 }
+          data:
+              failCommands: ["insert"]
+              closeConnection: true
+      - name: insertOne
+        object: collection
+        arguments:
+          session: session0
+          document:
+            _id: 4
+        result:
+          errorLabelsContain: ["TransientTransactionError"]
+          errorLabelsOmit: ["UnknownTransactionCommitResult"]
+      # Session remains pinned to the first mongos and commitTransaction
+      # succeeds.
+      - name: commitTransaction
+        object: session0
+
+    expectations:
+      - command_started_event:
+          command:
+            insert: *collection_name
+            documents:
+              - _id: 3
+            ordered: true
+            readConcern:
+            lsid: session0
+            txnNumber:
+              $numberLong: "1"
+            startTransaction: true
+            autocommit: false
+            writeConcern:
+          command_name: insert
+          database_name: *database_name
+      - command_started_event:
+          command:
+            insert: *collection_name
+            documents:
+              - _id: 4
+            ordered: true
+            readConcern:
+            lsid: session0
+            txnNumber:
+              $numberLong: "1"
+            startTransaction:
+            autocommit: false
+            writeConcern:
+          command_name: insert
+          database_name: *database_name
+      - command_started_event:
+          command:
+            commitTransaction: 1
+            lsid: session0
+            txnNumber:
+              $numberLong: "1"
+            startTransaction:
+            autocommit: false
+            writeConcern:
+            recoveryToken: 42
+          command_name: commitTransaction
+          database_name: admin
+
+    outcome:
+      collection:
+        data:
+          - _id: 1
+          - _id: 2
+          - _id: 3
+
+  - description: remain pinned after errors within a transaction and abort
```

Comment:
I would like to hear some thoughts one this behavior. As described in the spec (and tested here) a session is pinned to a mongos even after a network error while running a CRUD operation within the transaction. This has the interesting side effect that a subsequent `abort` attempt will block until the serverSelectionTimeout (assuming the operation failed because mongos is down/unreachable).

Since it will be common to retry the entire transaction in this case it seems somewhat pointless to remain pinned for the initial abort attempt. I did not attempt to workaround this case because I wanted to keep the session pin/unpin behavior simple. Any thoughts?

Reply:
Can you propose an alternative so we can evaluate?  I imagine it would involve some sort of storage of the last exception in the session and then basing the the abortTransaction behavior off of that?

Let's check with Aly on this as well.

Reply:
> I imagine it would involve some sort of storage of the last exception in the session and then basing the the abortTransaction behavior off of that?

My alternative proposal would be to unpin after transient errors within a transaction. This way a subsequent abort attempt would already be unpinned and could immediately select any available mongos. In this proposal a commit attempt that selects a new mongos would error but I wouldn't expect an app to commit after a transient error within the transaction.

Reply:
I like that better if it works

Reply:
I updated the prose to say that drivers MUST unpin a session after any transient error within a transaction. This ensures that a subsequent abort does not block and so a subsequent transaction (like withTransaction) can proceed immediately.

Note this new behavior comes with its own downside though. Due to server limitations in 4.2, an abort on a new mongos will not actually abort the in-progress transaction on the shards. The transaction will remain open and potentially block other operations. This case is possible when a mongos crashes regardless of driver behavior, but if the connection error was simply a blimp and the mongos is healthy the entire time unpinning will cause more transactions to be left open on the server.

Personally, I prefer the updated prose. @jmikola, @EshaMaharishi, thoughts?

Reply:
> This ensures that a subsequent abort does not block

Why would the driver send abort after hearing TransientTransactionError?

Reply:
In the case of a network error (which is labelled a TransientTransactionError) the driver needs to send abort to release the transaction's resources. When the server returns a command error including the "TransientTransactionError" label, I suppose drivers could skip sending abortTransaction because the server has already aborted the transaction. This is an optimization that we considered in 4.0 but rejected for the sake of simplicity.

Reply:
I'm confused, does the driver label network errors as TransientTransactionError itself?

Reply:
Yes. The driver adds the TransientTransactionError error label to network errors: https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#transienttransactionerror

> Any command error that includes the "TransientTransactionError" error label in the "errorLabels" field. Any network error or server selection error encountered running any command besides commitTransaction in a transaction. In the case of command errors, the server adds the label; in the case of network errors or server selection errors where the client receives no server reply, the client adds the label.

Reply:
I see. So if the driver gets a "blip" network error and unpins from the original mongos, the driver may send abort to a different mongos, and that is the motivation for https://jira.mongodb.org/browse/SERVER-39692?

Reply:
Exactly. Other similar cases are when the original mongos is restarted and looses the in memory transaction state or the original mongos is healthy but there's a network partition preventing any connections from the driver.

Reply:
Ok, got it. I have no further comments.